kotlinx-coroutines-core/common/src/CancellableContinuation.kt

```diff
@@ -4,31 +4,97 @@ import kotlinx.coroutines.internal.*
 import kotlin.coroutines.*
 import kotlin.coroutines.intrinsics.*
 
-// --------------- cancellable continuations ---------------
-
 /**
- * Cancellable continuation. It is _completed_ when resumed or cancelled.
- * When the [cancel] function is explicitly invoked, this continuation immediately resumes with a [CancellationException] or
+ * Cancellable [continuation][Continuation] is a thread-safe continuation primitive with the support of
+ * an asynchronous cancellation.
+ *
+ * Cancellable continuation can be [resumed][Continuation.resumeWith], but unlike regular [Continuation],
+ * it also might be [cancelled][CancellableContinuation.cancel] explicitly or [implicitly][Job.cancel] via a parent [job][Job].
+ *
+ * If the continuation is cancelled successfully, it resumes with a [CancellationException] or
  * the specified cancel cause.
  *
- * An instance of `CancellableContinuation` is created by the [suspendCancellableCoroutine] function.
+ * ### Usage
+ *
+ * An instance of `CancellableContinuation` can only be obtained by the [suspendCancellableCoroutine] function.
+ * The interface itself is public for use and private for implementation.
+ *
+ * A typical use of this function is to suspend a coroutine while waiting for a result
+ * from a callback or external source of values:
```

Comment:
```suggestion
 * A typical usage of this function is to suspend a coroutine while waiting for a result
 * from a callback or an external source of values that also support cancelling.
 * If the callback does not support cancelling, consider using [suspendCoroutine].
 *
 * Example:
```
I saw a question like, "What do I pass to `invokeOnCancellation` if my callback can't be unregistered?"

Reply:
>or an external source of values that also support cancelling.

This seems to be misleading -- the source might not support cancellation, but our suspending function does.
Using `suspendCoroutine` removes cancellation via Job completely. (Also, support**s** or you've meant something else?)

Reply:
Rephrased it, PTAL

Reply:
Cancelling the coroutine without cancelling the callback actually causes confusion (I'm not sure that I can find a link to one of the relevant discussions quickly, though): "Hey, I've cancelled the request almost immediately, how come it still goes through?" There's clearly some expectation that coroutine cancellation is magic that automatically deals with blocking calls and unregistering callbacks. We can't do much about this, but at least the docs should expose the brutal truth to those who actually read them.

Don't let that block the PR, though. If you agree this is worth including and can think of correct phrasing that highlights that it's fully up to the programmer to integrate cancellation / what happens when cancellation is only supported partially, please go ahead; if not, we can look into this later.